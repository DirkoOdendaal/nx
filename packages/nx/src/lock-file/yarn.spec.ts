import {
  parseYarnLockFile,
  pruneYarnLockFile,
  stringifyYarnLockFile,
} from './yarn';
import {
  berryLockFile,
  berryLockFileDevkitAndYargs,
  berryRxjsTslibLockFile,
  berrySsh2LockFile,
  lockFile,
  lockFileDevkitAndYargs,
  lockFileJustTypescript,
  rxjsTslibLockFile,
  ssh2LockFile,
} from './__fixtures__/yarn.lock';

const TypeScriptOnlyPackage = {
  name: 'test',
  version: '1.0.0',
  dependencies: { typescript: '4.8.4' },
};
const YargsAndDevkitPackage = {
  name: 'test',
  version: '0.0.0',
  dependencies: {
    '@nrwl/devkit': '15.0.13',
    yargs: '17.6.2',
    typescript: '4.8.4',
  },
};
const YargsDevkitTypescriptPackage = {
  name: 'test',
  version: '0.0.0',
  dependencies: {
    '@nrwl/devkit': '15.0.13',
    typescript: '4.8.4',
    yargs: '17.6.2',
  },
};
const Ssh2Package = {
  name: 'test',
  version: '0.0.0',
  dependencies: {
    ssh2: '1.11.0',
  },
};
const RxjsTslibPackage = {
  name: 'test',
  version: '0.0.0',
  dependencies: {
    rxjs: '^7.8.0',
    tslib: '^2.4.1',
  },
};

describe('yarn LockFile utility', () => {
  describe('classic', () => {
    const parsedLockFile = parseYarnLockFile(lockFile);

    it('should parse lockfile correctly', () => {
      expect(parsedLockFile.lockFileMetadata).toBeUndefined();
      expect(Object.keys(parsedLockFile.dependencies).length).toEqual(339);
      expect(
        parsedLockFile.dependencies['@ampproject/remapping']
      ).toMatchSnapshot();
      expect(parsedLockFile.dependencies['typescript']).toMatchSnapshot();
    });

    it('should map various versions of packages', () => {
      expect(
        Object.keys(parsedLockFile.dependencies['@jridgewell/gen-mapping'])
          .length
      ).toEqual(2);
      expect(
        parsedLockFile.dependencies['@jridgewell/gen-mapping'][
          '@jridgewell/gen-mapping@0.1.1'
        ]
      ).toBeDefined();
      expect(
        parsedLockFile.dependencies['@jridgewell/gen-mapping'][
          '@jridgewell/gen-mapping@0.1.1'
        ].rootVersion
      ).toBeFalsy();
      expect(
        parsedLockFile.dependencies['@jridgewell/gen-mapping'][
          '@jridgewell/gen-mapping@0.3.2'
        ]
      ).toBeDefined();
      expect(
        parsedLockFile.dependencies['@jridgewell/gen-mapping'][
          '@jridgewell/gen-mapping@0.3.2'
        ].rootVersion
      ).toBeTruthy();
    });

    it('should map various instances of the same version', () => {
      const babelCoreDependency =
        parsedLockFile.dependencies['@babel/core']['@babel/core@7.20.5'];
      expect(babelCoreDependency.packageMeta.length).toEqual(2);
      expect(babelCoreDependency.packageMeta).toEqual([
        '@babel/core@^7.11.6',
        '@babel/core@^7.12.3',
      ]);
    });

    it('should match the original file on stringification', () => {
      expect(stringifyYarnLockFile(parsedLockFile)).toEqual(lockFile);
    });

    it('should prune the lock file', () => {
      expect(
        Object.keys(
          pruneYarnLockFile(parsedLockFile, TypeScriptOnlyPackage).dependencies
        ).length
      ).toEqual(1);
      expect(
        Object.keys(
          pruneYarnLockFile(parsedLockFile, YargsAndDevkitPackage).dependencies
        ).length
      ).toEqual(37);
    });

    it('should correctly prune lockfile with single package', () => {
      expect(
        stringifyYarnLockFile(
          pruneYarnLockFile(parsedLockFile, TypeScriptOnlyPackage)
        )
      ).toEqual(lockFileJustTypescript);
    });

    it('should correctly prune lockfile with multiple packages', () => {
      expect(
        stringifyYarnLockFile(
          pruneYarnLockFile(parsedLockFile, YargsAndDevkitPackage)
        )
      ).toEqual(lockFileDevkitAndYargs);
    });

    it('should correctly prune lockfile with package that has optional dependencies', () => {
      expect(
        stringifyYarnLockFile(
          pruneYarnLockFile(parseYarnLockFile(ssh2LockFile), Ssh2Package)
        )
      ).toEqual(ssh2LockFile);
    });

    it('should correctly prune lockfile with packages in multiple versions', () => {
      expect(
        stringifyYarnLockFile(
          pruneYarnLockFile(
            parseYarnLockFile(rxjsTslibLockFile),
            RxjsTslibPackage
          )
        )
      ).toEqual(rxjsTslibLockFile);
    });
  });

  describe('berry', () => {
    const parsedLockFile = parseYarnLockFile(berryLockFile);

    it('should parse lockfile correctly', () => {
      expect(parsedLockFile.lockFileMetadata).toEqual({
        __metadata: { cacheKey: '8', version: '6' },
        workspacePackages: {
          'test@workspace:.': {
            dependencies: {
              '@nrwl/cli': '15.0.13',
              '@nrwl/workspace': '15.0.13',
              nx: '15.0.13',
              prettier: '^2.6.2',
              typescript: '~4.8.2',
            },
            languageName: 'unknown',
            linkType: 'soft',
            resolution: 'test@workspace:.',
            version: '0.0.0-use.local',
          },
        },
      });
      expect(Object.keys(parsedLockFile.dependencies).length).toEqual(401);
      expect(
        parsedLockFile.dependencies['@ampproject/remapping']
      ).toMatchSnapshot();
      expect(parsedLockFile.dependencies['typescript']).toMatchSnapshot();
    });

    it('should map various versions of packages', () => {
      expect(
        Object.keys(parsedLockFile.dependencies['@jridgewell/gen-mapping'])
          .length
      ).toEqual(2);
      expect(
        parsedLockFile.dependencies['@jridgewell/gen-mapping'][
          '@jridgewell/gen-mapping@0.1.1'
        ]
      ).toBeDefined();
      expect(
        parsedLockFile.dependencies['@jridgewell/gen-mapping'][
          '@jridgewell/gen-mapping@0.3.2'
        ]
      ).toBeDefined();
    });

    it('should map various instances of the same version', () => {
      const babelCoreDependency =
        parsedLockFile.dependencies['@babel/core']['@babel/core@7.20.5'];
      expect(babelCoreDependency.packageMeta.length).toEqual(2);
      expect(babelCoreDependency.packageMeta).toEqual([
        '@babel/core@npm:^7.11.6',
        '@babel/core@npm:^7.12.3',
      ]);
    });

    it('should match the original file on stringification', () => {
      const result = stringifyYarnLockFile(parsedLockFile);
      expect(result).toMatch(
        /This file was generated by Nx. Do not edit this file directly/
      );
      expect(removeComment(result)).toEqual(removeComment(berryLockFile));
    });

    it('should prune the lock file', () => {
      expect(
        Object.keys(
          pruneYarnLockFile(parsedLockFile, YargsDevkitTypescriptPackage)
            .dependencies
        ).length
      ).toEqual(38);
    });

    it('should correctly prune lockfile with multiple packages', () => {
      const result = stringifyYarnLockFile(
        pruneYarnLockFile(parsedLockFile, YargsDevkitTypescriptPackage)
      );
      expect(removeComment(result)).toEqual(
        removeComment(berryLockFileDevkitAndYargs)
      );
    });

    it('should correctly prune lockfile with multiple packages and custom name', () => {
      const result = pruneYarnLockFile(parsedLockFile, {
        ...YargsDevkitTypescriptPackage,
        name: 'custom-name',
      });
      expect(result.lockFileMetadata.workspacePackages).toMatchInlineSnapshot(`
        Object {
          "custom-name@workspace:^": Object {
            "dependencies": Object {
              "@nrwl/devkit": "15.0.13",
              "typescript": "4.8.4",
              "yargs": "17.6.2",
            },
            "languageName": "unknown",
            "linkType": "soft",
            "resolution": "custom-name@workspace:^",
            "version": "0.0.0-use.local",
          },
        }
      `);
    });

    it('should correctly prune lockfile with package that has optional dependencies', () => {
      expect(
        removeComment(
          stringifyYarnLockFile(
            pruneYarnLockFile(parseYarnLockFile(berrySsh2LockFile), Ssh2Package)
          )
        )
      ).toEqual(removeComment(berrySsh2LockFile));
    });

    it('should correctly prune lockfile with packages in multiple versions', () => {
      expect(
        removeComment(
          stringifyYarnLockFile(
            pruneYarnLockFile(
              parseYarnLockFile(berryRxjsTslibLockFile),
              RxjsTslibPackage
            )
          )
        )
      ).toEqual(removeComment(berryRxjsTslibLockFile));
    });
  });
});

// we don't care about comment message
const removeComment = (value) => value.split(/\n/).slice(2).join('\n');
