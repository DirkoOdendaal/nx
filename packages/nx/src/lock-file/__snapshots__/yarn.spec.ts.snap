// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`yarn LockFile utility berry should parse lockfile correctly 1`] = `
Object {
  "@ampproject/remapping@2.2.0": Object {
    "checksum": "d74d170d06468913921d72430259424b7e4c826b5a7d39ff839a29d547efb97dc577caa8ba3fb5cf023624e9af9d09651afc3d4112a45e2050328abc9b3a2292",
    "dependencies": Object {
      "@jridgewell/gen-mapping": "^0.1.0",
      "@jridgewell/trace-mapping": "^0.3.9",
    },
    "languageName": "node",
    "linkType": "hard",
    "packageMeta": Array [
      "@ampproject/remapping@npm:^2.1.0",
    ],
    "resolution": "@ampproject/remapping@npm:2.2.0",
    "rootVersion": true,
    "version": "2.2.0",
  },
}
`;

exports[`yarn LockFile utility berry should parse lockfile correctly 2`] = `
Object {
  "typescript@4.8.4": Object {
    "bin": Object {
      "tsc": "bin/tsc",
      "tsserver": "bin/tsserver",
    },
    "checksum": "3e4f061658e0c8f36c820802fa809e0fd812b85687a9a2f5430bc3d0368e37d1c9605c3ce9b39df9a05af2ece67b1d844f9f6ea8ff42819f13bcb80f85629af0",
    "languageName": "node",
    "linkType": "hard",
    "packageMeta": Array [
      "typescript@npm:~4.8.2",
    ],
    "resolution": "typescript@npm:4.8.4",
    "rootVersion": true,
    "version": "4.8.4",
  },
}
`;

exports[`yarn LockFile utility classic should parse lockfile correctly 1`] = `
Object {
  "@ampproject/remapping@2.2.0": Object {
    "dependencies": Object {
      "@jridgewell/gen-mapping": "^0.1.0",
      "@jridgewell/trace-mapping": "^0.3.9",
    },
    "integrity": "sha512-qRmjj8nj9qmLTQXXmaR1cck3UXSRMPrbsLJAasZpF+t3riI71BXed5ebIOYwQntykeZuhjsdweEc9BxH5Jc26w==",
    "packageMeta": Array [
      "@ampproject/remapping@^2.1.0",
    ],
    "resolved": "https://registry.yarnpkg.com/@ampproject/remapping/-/remapping-2.2.0.tgz#56c133824780de3174aed5ab6834f3026790154d",
    "rootVersion": true,
    "version": "2.2.0",
  },
}
`;

exports[`yarn LockFile utility classic should parse lockfile correctly 2`] = `
Object {
  "typescript@4.8.4": Object {
    "integrity": "sha512-QCh+85mCy+h0IGff8r5XWzOVSbBO+KfeYrMQh7NJ58QujwcE22u+NUSmUxqF+un70P9GXKxa2HCNiTTMJknyjQ==",
    "packageMeta": Array [
      "typescript@~4.8.2",
    ],
    "resolved": "https://registry.yarnpkg.com/typescript/-/typescript-4.8.4.tgz#c464abca159669597be5f96b8943500b238e60e6",
    "rootVersion": true,
    "version": "4.8.4",
  },
}
`;
